File: public/dsa/linked_lists/kth_element_in_linked_list/approach/kth_element_in_linked_list.cpp

#include <bits/stdc++.h>
using namespace std;

// defs-start
class ListNode {
   public:
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};
// defs-end

ListNode* kthElement(ListNode* head, int k) {
    ListNode* temp = head;

    // Move (k - 1) steps ahead to reach the kth node
    int count = 1;
    while (temp != NULL && count < k) {
        temp = temp->next;
        count++;
    }

    return temp;
}

// tests-start
int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(3);
    head->next->next = new ListNode(4);
    head->next->next->next = new ListNode(7);
    int k = 2;
    auto node = kthElement(head, k);
    cout << node->data << endl;
    return 0;
}
// tests-end

/*output
3
*/

File: public/dsa/linked_lists/kth_element_in_linked_list/approach/kth_element_in_linked_list.py
from typing import List


# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def kthElement(self, head: ListNode, k: int) -> ListNode:
        temp = head
        count = 1

        # Move (k - 1) steps ahead to reach the kth node
        while temp is not None and count < k:
            temp = temp.next
            count += 1

        return temp


# tests-start
head = ListNode(1, ListNode(3, ListNode(4, ListNode(7))))
k = 2
sol = Solution()
node = sol.kthElement(head, k)
print(node.data)
# tests-end

'''output
3
'''
